Add support for GitHub Enterprise

mdwheele opened the pull request:
Related: #857, #240

This is a follow up to #857 that addresses the review given by @tarebyte. Since it's been a while, it's probably best to start with resolution to all of those and then give another review. I have rebased my work onto `master` and have yet to set up the development environment again, so tests will likely fail. Gonna give Docker a go to get all of this going again.

mdwheele commented:
@tarebyte I think I've implemented the changes you requested in the previous review as well as covered as much of hound's style violations. The status check is still failing, but I don't know what the "2 violations found" are. 

I'm also unsure how to approach the spec failures so advice would be appreciated there.

mdwheele commented:
Looks like [the tests that failed in this build](https://travis-ci.org/education/classroom/jobs/278880600) were not expecting `:auto_paginate` from the client. I turned it off globally and [it passed](https://travis-ci.org/education/classroom/jobs/278882297). There are a lot of callers of `github_client` that pass in this option when needed. I think the failures I have are a symptom of creating a single instance of the client with one configuration.

My feeling is that we can either disable auto pagination globally and then per-caller, add it where needed. Alternatively, we can enable it globally and adjust VCR files because I really think it's just bawking at getting a `&per_page=100` in the URI that it doesn't expect. I'll verify that and report back.

mdwheele commented on lib/github/search.rb:
```diff
@@ -7,7 +7,7 @@ class Search
       options[:access_token] = access_token
 
-      @search_client = Octokit::Client.new(options)
+      @search_client = GitHubClassroom.github_client(options)
```
Yeah, check out this method. auto_paginate is passed in through options and if we do a reassignment of that within GitHubClassroom.github_client, we're changing behaviour inadvertently. I'll push one more commit that makes it so the factory for the client doesn't make a decision on pagination. It defers to whatever the caller passes in.

jeffrafter replied:
The change to the factory for this comment ☝️ was added... however there was a small side-effect: existing consumers of the factory method had their `auto_paginate` behavior changed (which previously defaulted to true). 

These are the three impacted call sites:

https://github.com/mdwheele/classroom/blob/281877f6a0d636fe1413d7f69deaf102956dc8f9/app/models/github_model.rb#L181
https://github.com/mdwheele/classroom/blob/281877f6a0d636fe1413d7f69deaf102956dc8f9/app/models/github_user.rb#L16
https://github.com/mdwheele/classroom/blob/281877f6a0d636fe1413d7f69deaf102956dc8f9/lib/github/token.rb#L46

jeffrafter replied:
I think all of these are fixed now! 🎉

mdwheele commented:
Cool. Looks like I got everything passing. It turns out that when I was merging the previous `github_client` factory with the suggested changes, I introduced a bug. The way the `tap` was written was wiping out `auto_paginate`.

I figured out how what hound was checking against and ran Rubocop to fix up the last style violations.

Sorry for all the noise. I blame hound.

mdwheele commented:
Hey y'all! Is there an update on this or anything I can do?

tarebyte commented:
@mdwheele my apologies, it's just me maintaining the application right now and I've been swamped 😄 

I hope to get a look at this early next week, thanks for your patience!

mdwheele commented:
Hey @tarebyte! No worries, I understand. Not trying to rush, but can I get an indication of where this lies; priority-wise? I only ask as I have folks on my end asking about the feature. Thanks! Hope all is well.

d12 commented on app/views/assignments/_assignment_form_options.html.erb:
```diff
@@ -35,12 +35,15 @@
   <dd class="dd-autocomplete-suggestions"><%= render_autocomplete_suggestions_container %></dd>
   <input class="js-autocomplete-resource-id" type="hidden" name="repo_id" />
 </dl>
+<% end %>
```
Why is starter code not an option for enterprise users?

mdwheele replied:
See the [conversation from #240](https://github.com/education/classroom/issues/240#issuecomment-274623526). In short, the Enterprise API does not support the Source Import API. I talked with our campus representative and there was no clear ETA on that feature either.

d12 replied:
That's too bad :/ Oh well, this is fine for now.

jeffrafter replied:
With repo-templates I think we will be able to reintroduce this. But for now I think this is a good trade off.

d12 commented on app/views/assignments/new.html.erb:
```diff
@@ -18,7 +18,9 @@
       </div>
 
+      <% if !GitHubClassroom.enterprise_instance? %>
       <%= render partial: 'shared/unlimited_free_private_repositories_banner' %>
```
Can you indent this line by one?

d12 commented on config/initializers/github_client.rb:
```diff
@@ -2,12 +2,22 @@
+      unless options.key?(:access_token)
+        opts[:client_id]     = Rails.application.secrets.github_client_id
+        opts[:client_secret] = Rails.application.secrets.github_client_secret
+      end
```
The old version of `github_client` would pass `client_id`, `client_secret`, and `access_token` if `access_token` was provided. Any reason for the change here?

mdwheele replied:
My understanding is that the current code-base uses client_id and client_secret to establish an access_token through the OAuth2 Auth Code Grant. If an access_token is provided, there seems to be no reason to pass in a client_id and client_secret. Currently, this happens as a side-effect of always passing in client_id and client_secret regardless.

If Classroom were using the Client Credentials grant for anything against GitHub.com or GitHub Enterprise, then it makes sense to configure those attributes. See https://github.com/octokit/octokit.rb#oauth-access-tokens and https://github.com/octokit/octokit.rb#application-authentication.

Does Classroom use the Client Credentials grant for any operations, currently?

d12 commented on config/initializers/github_client.rb:
```diff
@@ -2,12 +2,22 @@
-      auto_paginate: true
-    }.merge!(options)
+    client_options = options.tap do |opts|
+      opts[:user_agent] = "GitHub Classroom"
```
Why do we need to set this now?

mdwheele replied:
See [this comment from @tarebyte](https://github.com/education/classroom/pull/857#discussion_r110309411) from my previous PR on this issue.

mdwheele commented:
Update from https://github.com/mdwheele/classroom/issues/1

There is, indeed, a bug that's surfaced in the time between this PR being opened and now, it seems.

The Collaborators API endpoint is returning 204 No Content on Enterprise and the correct 200 OK with contents on GitHub.com (see https://developer.github.com/enterprise/2.11/v3/repos/collaborators/#response-1).

I'm going to implement a workaround so we can get further and at the same time reach out to our representative with GitHub and see what the deal is. The feature is documented in the enterprise API documentation as "there".

d12 commented:
@mdwheele That endpoint returns 204 if the user is already a collaborator on the repository. Is that the case in your example? If not, I can dig into this more. Any other relavent information would be helpful.

mdwheele commented:
Sure thing.

I'm doing this just by `curl` against the API; no GitHub Classroom involvement. I have a repository named `mdwheele-testing` and I removed all collaborators. Interestingly, when I query for collaborators, I still come back in results (as you'll see in output). I added a different account and still received `204 No Content`:

```
mdwheele@blaze:/usr/src/app$ curl -v -u [redacted]:[redacted] -X GET -H "Accept: application/vnd.github.hellcat-preview+json" -H "Content-Length: 0" https://[redacted]/api/v3/repositories/[redacted]/collaborators
> GET /api/v3/repositories/[redacted]/collaborators HTTP/1.1
> Authorization: Basic [redacted]
> User-Agent: curl/7.38.0
> Host: [redacted]
> Accept: application/vnd.github.hellcat-preview+json
> Content-Length: 0
> 
< HTTP/1.1 200 OK
* Server GitHub.com is not blacklisted
< Server: GitHub.com
< Date: Wed, 29 Nov 2017 22:17:11 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 1211
< Status: 200 OK
< Cache-Control: private, max-age=60, s-maxage=60
< Vary: Accept, Authorization, Cookie, X-GitHub-OTP
< ETag: "e28fd260f9bf41bcee5a097a1ebab0d2"
< X-OAuth-Scopes: repo, user
< X-Accepted-OAuth-Scopes: 
< X-GitHub-Media-Type: github.hellcat-preview; format=json
< Access-Control-Expose-Headers: ETag, Link, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval
< Access-Control-Allow-Origin: *
< X-GitHub-Request-Id: 18155b34-bb58-48fb-a138-337b8df311a9
< Content-Security-Policy: default-src 'none'
< Strict-Transport-Security: max-age=31536000; includeSubdomains
< X-Content-Type-Options: nosniff
< X-Frame-Options: deny
< X-XSS-Protection: 1; mode=block
< X-Runtime-rack: 0.071847
< 
[
  {
    "login": "[redacted]",
    "id": [redacted],
    "avatar_url": "https://avatars.[redacted]/u/[redacted]?",
    "gravatar_id": "",
    [snip]
    "type": "User",
    "site_admin": false,
    "ldap_dn": "[redacted]",
    "permissions": {
      "admin": true,
      "push": true,
      "pull": true
    }
  }
]
* Connection #0 to host [redacted] left intact

mdwheele@blaze:/usr/src/app$ curl -v -u [redacted]:[redacted] -X PUT -H "Accept: application/vnd.github.hellcat-preview+json" -H "Content-Length: 0" https://github.ncsu.edu/api/v3/repositories/[redacted]/collaborators/[other-user]
> PUT /api/v3/repositories/[redacted]/collaborators/[other-user] HTTP/1.1
> Authorization: Basic [redacted]
> User-Agent: curl/7.38.0
> Host: [redacted]
> Accept: application/vnd.github.hellcat-preview+json
> Content-Length: 0
> 
< HTTP/1.1 204 No Content
* Server GitHub.com is not blacklisted
< Server: GitHub.com
< Date: Wed, 29 Nov 2017 22:18:03 GMT
< Status: 204 No Content
< X-OAuth-Scopes: repo, user
< X-Accepted-OAuth-Scopes: 
< X-GitHub-Media-Type: github.hellcat-preview; format=json
< Access-Control-Expose-Headers: ETag, Link, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval
< Access-Control-Allow-Origin: *
< X-GitHub-Request-Id: b169ee09-5bfc-440b-a20b-60865d1d3259
< Content-Security-Policy: default-src 'none'
< Strict-Transport-Security: max-age=31536000; includeSubdomains
< X-Content-Type-Options: nosniff
< X-Frame-Options: deny
< X-XSS-Protection: 1; mode=block
< X-Runtime-rack: 0.143426
< 
* Connection #0 to host [redacted] left intact
```

mdwheele commented:
I'll hold until I hear back @d12. Just let me know. This workaround would actually have to go to Octokit, I think. I may hack it *just* to make sure this is an isolated issue. I doubt a one-off for Enterprise would be merged into Octokit without a larger-scope "compatability shim" that bridges differences between GitHub and GitHub Enterprise. But who knows.

d12 commented:
I've made an internal issue, we'll check this out when we can.

mdwheele commented:
Excellent. Worth mentioning that we're running 2.11.2 but I haven't seen anything in the notes since then. The other user that reported an issue with Thisbe branch was some patch of 2.9

mdwheele commented:
@d12 Have there been any updates to this issue? AFAICT, the bug is still present in 2.12.3.

d12 commented:
Looking at the documentation, it seems 204 is the correct response: 

https://developer.github.com/v3/repos/collaborators/#response-1
https://developer.github.com/enterprise/2.12/v3/repos/collaborators/#response-1

I don't believe this endpoint returns a 200.

mdwheele commented:
Octokit expects a 201 response, not 200. It sends the preview media type, which is documented but does not work, as documented, on GutHub Enterprise 2.12.3

d12 commented:
Hey @mdwheele sorry about the late response. We have a fix in PR for this right now, but I don't know when exactly it will ship. Once it ships, this endpoint will return the 201 that Octokit is expecting.

I'll ping you here when this ships.

trombonekenny commented:
Ping pong! Any progress on the endpoint's PR to return 201? Thanks.

mdwheele commented:
We're planning on doing the 2.13 upgrade very soon and I'll be giving this another shot. Maybe @d12 has some more information on delivery of that PR. 2.13 is so shiny! :tada:

mdwheele commented:
@d12 

Looks like it is fixed. We're currently running 2.12.8 (about to upgrade to 2.13). I will pick this PR back up soon as there are a few other items to take care of (mostly visual). Enough time has passed that it's likely a good idea to rebase and go from there. Thanks for all the help so far!

d12 commented:
Great, thanks so much for working on this, and sorry for the delays!

mdwheele commented:
VOILA! Seems to work excellently for us. We're running 2.12.8 at the moment and at 5:30pm we will hopefully be running 2.13.1. I'll give this another test tomorrow on 2.13 for giggles, but I really thing this is good to go now.

I also caught the fact that we were not hiding the starter code form for group assignments (See 6aa33bd). 

If you would like me to rebase this onto `upstream/master`, I can do that to clean up the commit history. 

@d12 @tarebyte 

Do either of  you know whether or not the source import API will ever be added to GitHub Enterprise? I know there are several folks on my end that do want this feature, but are willing to pass on it to use this application at all. If amenable, I would also be willing to contribute a maintainable option that would work for both.

d12 commented:
Thanks for taking care of this @mdwheele !

We don't really have a ton of bandwidth to review this, but we'll try to check this out as soon as we can. One thing we're still unsure about is how to test this, to make sure all of the functionality works for a enterprise GitHub API. We haven't given it too much focused though yet, but I'm curious if you had any ideas re: that.

mdwheele commented:
> We haven't given it too much focused though yet, but I'm curious if you had any ideas re: that.

Sure! If there is value in full end-to-end tests, we could create (and have to maintain) separate VCR cassettes for Enterprise API interactions (which, for the most part are the same, until they're different; as we've seen here). This seems like a lot of overhead and will require the project maintaining a test instance of GitHub Enterprise as part of the CI process (somewhere). Alternatively, the project could rely on the community of Enterprise users keeping that part of the suite "honest". It **is** in our interest to do so, after all.

That said...

As a low-tech alternative, we could evaluate adding / improving tests around our integration with OctoKit. OctoKit has coverage over GitHub.com and GitHub Enterprise. As long as this projects interface to "a GitHub of any flavor" is through OctoKit, then I don't know how valuable end-to-end tests for GitHub Enterprise are to Classroom. We could simply specify assertions against where we interface with OctoKit and stop there; relying on the OctoKit test-suite to provide adequate coverage for Enterprise.

Thoughts?

P.S. No worries on time-to-review. With this branch in place, I (and others that have been pinging me separately over the months) have enough to move forward; if even for trials / evaluations of the tool for their faculty, students and staff. Thanks for **your** time so far!

hjmjohnson commented:
@mdwheele Thank you for continuing to work on this important effort.

mdwheele commented:
Hey everyone; just checking in to see where this lies. We're piloting use of Classroom at NC State and would prefer to rely on an upstream branch.

Hope all is well!

hjmjohnson commented:
I've been following this PR for a long time now.  I'm very sad that this seems to have hit a road block .

d12 commented:
Unfortunately we haven't discovered a complete testing solution for this yet. 

One option would be to leave the API bits untested for now, and release this under preview/alpha for enterprise. All the application logic is still tested, it'll just be the interactions with the GitHub API.

Let me revisit this when I have a little bit more time, maybe the end of this week. We'll see what we can do :)

PReimers commented:
Hey @mdwheele, @d12 

Any news here?
As far as I can see there are still merge conflicts.

We would be pleased to test this if there would be a preview/alpha for this.

mdwheele commented:
Hey @PReimers! I'm still down to resolve conflicts and get this branch current with master. However, I'd like some indication that the work would be on the path to merging before doing so.

mdwheele commented:
Pokey poke poke.

PReimers commented:
@mdwheele So if I've got it right everything depends on @d12?

If it's going to be merged, the conflicts will be solved by you, if not we're waiting until they close this PR.

mdwheele commented:
@PReimers @d12 I have rebased this work onto `upstream/master`. I'll make some time to test against GitHub Enterprise 2.16.1 later today and will report back.

d12 commented:
Hey all :wave:

Unfortunately I don't have time to help out with this, I've been busy with other work lately. @BenEmdon and @srinjoym are our two engineers on Classroom at the moment, they may be able to help out with getting this 🚢 'd

BenEmdon commented:
Unfortunately I am also low on bandwidth at the moment and I imagine the review could be quite involved so I will defer to @srinjoym to guide you through the review process. I'll be around for a second approval.

mdwheele commented:
Hey again! The review for this is actually quite small. It's a feature toggle and reconfiguration of Octocat with a refactor to make sure every caller uses a single client reference. Previously, several client configs we're floating about. 

For features unsupported by the Enterprise API, we disable the web interface. Source code import is not available on Enterprise so we hide the UI for starter code as a compromise. 

Currently, I have several faculty shipping renditions of scripts back and forth and hacking together what Classroom provides effectively. I would like to contribute to this effort rather than building a self-service solution for them myself.

srinjoym commented on config/initializers/omniauth.rb:
```diff
@@ -1,8 +1,19 @@
     Rails.application.secrets.github_client_id,
     Rails.application.secrets.github_client_secret,
-    scope: "user:email,repo,delete_repo,admin:org,admin:org_hook"
+    { client_options: options }
```
This block (starting from `Rails.application.config.middleware.use OmniAuth:: Builder do`) needs to either be moved outside of the `{}.tap` call or the variable passed into `client_options`needs to be `opts` not `options`. Since `options` is initialized after the tap block, this will always pass nil for `client_options`

srinjoym commented on config/initializers/github_client.rb:
```diff
@@ -2,12 +2,22 @@
     Octokit::Client.new(client_options)
   end
+
+  def self.enterprise_instance?
```
Is there any reason we couldn't just use a feature flag to enable enterprise support instead of adding another secret? That would follow the pattern of our other new features and would make it easier to toggle this

mdwheele replied:
Definitely! Could you explain this a bit better? Currently, I've added two environment variables (See `.env.example` / `ENTERPRISE_ENABLED`) and just following what I saw other code doing at the time, it looked like the way to get environment variables down to Rails was through secrets. 

If there's a way to query `ENTERPRISE_ENABLED` inside of this method, I'm 100% down to get rid of the secrets and just read from `ENV`.

srinjoym replied:
I was thinking we could add a feature flag in https://github.com/mdwheele/classroom/blob/feature/support-ghe/app/controllers/application_controller/feature_flags_dependency.rb to check if we want to run on an enterprise instance.

jeffrafter replied:
I've refactored this a bit... now the feature is enabled based on whether or not the hostname is present in the environment. In order for the feature to work we need that hostname in the env and because of this I think it makes sense to hold off on introducing a feature flag.

PReimers commented:
@mdwheele any new here?

d12 commented:
Hey all, a couple of updates GitHub Classroom + GitHub Enterprise:

I want to prioritize getting this work across the line, hopefully sometime in the next month or so if possible. A few things have changed,

1) GitHub now offers GitHub Enterprise to schools for free. https://education.github.com/schools. We don't want Classroom to be incompatible with the GitHub that more and more schools are beginning to use.

2) For the first time (ever), Classroom has a full time dedicated engineering team, and a dedicated Program Manager. This will make shipping larger, more ambitious projects like this more possible than it was before with only part time help from me and Mark + summer interns.

That being said, @mdwheele what's the current state of this PR? It'd be amazing if you could get this green, but if not, we can take it from here. Thanks for all the work you've done here 💖

mdwheele commented:
Hey @d12! 

Very glad to hear the project is going to get some more support. That's awesome!

I'm more than happy to resolve conflicts and get this back on the path to merge-able. I can make some time tomorrow to take care of this.

Thanks again.

jeffrafter commented on .env-example:
```diff
@@ -1,3 +1,6 @@
+ENTERPRISE_ENABLED=false
```
There seems to be a mismatch between this name and the one inside of `config/secrets.yml`:

```suggestion
GITHUB_ENTERPRISE_ENABLED=false
```

jeffrafter replied:
Instead, I think this might be better just detecting if a GITHUB_ENTERPRISE_HOSTNAME is present.

jeffrafter commented on config/initializers/github_client.rb:
```diff
@@ -2,12 +2,22 @@
   end
+
+  def self.enterprise_instance?
+    Rails.application.secrets.enterprise_enabled == true
```
I think this should be `github_enterprise_enabled`, additionally we should verify the data type here (boolean versus string):

```suggestion
    Rails.application.secrets.github_enterprise_enabled == true
```

jeffrafter replied:
Instead, I think this might be better just detecting if a `GITHUB_ENTERPRISE_HOSTNAME` is present.

jeffrafter commented on config/secrets.yml:
```diff
@@ -13,12 +13,14 @@
+  api_secret: <%= ENV['API_SECRET'] %>
+  google_client_id: <%= ENV['GOOGLE_CLIENT_ID'] %>
   google_client_secret: <%= ENV['GOOGLE_CLIENT_SECRET'] %>
+  github_enterprise_enabled: <%= ENV['GITHUB_ENTERPRISE'] %>
```
Related to my earlier `.env` comment above, we might want to change this:

```suggestion
  github_enterprise_enabled: <%= ENV['GITHUB_ENTERPRISE_ENABLED'] == 'true' %>
```

jeffrafter replied:
I think we should remove this variable (see above)

jeffrafter commented:
@mdwheele and others: I'm working on getting this up to date with all of the latest changes so that we can work to get these changes landed. I've been working on adding some coverage around the changes to `auto_paginate` to fully understand them. Thanks so much for the patience (2 years!)

favonia commented on config/initializers/omniauth.rb:
```diff
@@ -7,10 +7,19 @@ module Strategies
+    options[:site]          = "https://#{hostname}/api/v3"
+    options[:authorize_url] = "https://#{hostname}/login/oauth/authorize"
+    options[:token_url]     = "https://#{hostname}/login/oauth/access_token"
+  end
```
These three properties appear under `client_options` now. (I already connected GitHub Classroom to GitHub Enterprise in our school.)

jeffrafter replied:
Sorry, I'm not quite sure where you mean... do you mean in `github_client`? I can't find where these would be setup outside of this middleware.

favonia replied:
No, this is what I meant:
```ruby
options[:client_options] = {
  :site => "https://#{hostname}/api/v3",
  :authorize_url => "https://#{hostname}/login/oauth/authorize",
  :token_url => "https://#{hostname}/login/oauth/access_token",
}
```
See https://github.com/omniauth/omniauth-github.

mdwheele commented:
@jeffrafter Thanks for taking this on! I agree with everything you've done and we're excited to see this on a trajectory to be merged.

jeffrafter commented:
👋 

I think this is ready for a re-review. Using @mdwheele's work as a foundation and all of the comments in this pull request has been incredibly helpful. I wanted to call out a couple of goals:

- We should always be using the `github_client` factory to create instances of `Octokit`. This allows us to control the URL we are connecting to in a single location
- We should be explicitly supplying `auto_paginate` where it was previously being used - no pagination behavior should have changed
- We should allow for different URL patterns for GitHub Enterprise (including ports and proxies)
- Anywhere that we previously hard-linked to `github.com` should be configurable where the relevant information (like to an Organization) should live in the GitHub Enterprise instance
- You should be able to run a local version of GitHub Classroom (yay open source!)
- We should hide behavior not currently available on GitHub Enterprise (such as porter and repository templates)

jeffrafter commented:
Thanks @mdwheele for starting this process and being patient with it for so long (two years!). Thanks also to all of the folks reviewing, catching things and making comments, and helping test this. 💖
